File: MomentsJava/src/main/java/com/example/momentsjava/data/api/ListApiClient.java
package com.example.momentsjava.data.api;

import android.util.Log;

import com.example.momentsjava.data.model.ListItem;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.io.InputStreamReader;
import java.lang.reflect.Type;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.List;

import io.reactivex.rxjava3.core.Observable;


public class ListApiClient {

    private static final String DOMAIN_URL = "10.205.19.77";
    private static final int TIMEOUT = 10000;

    public Observable<List<ListItem>> getList() {
        return Observable.create(emitter -> {
            HttpURLConnection connection = null;
            try {
                connection = (HttpURLConnection) new URL("http://" + DOMAIN_URL + ":3022/list").openConnection();
                connection.setRequestMethod("GET");
                connection.setConnectTimeout(TIMEOUT);
                connection.setReadTimeout(10000);

                int responseCode = connection.getResponseCode();
                if (responseCode == HttpURLConnection.HTTP_OK) {
                    InputStreamReader reader = new InputStreamReader(connection.getInputStream());
                    StringBuilder response = new StringBuilder();
                    int data;
                    while ((data = reader.read()) != -1) {
                        response.append((char) data);
                    }

                    List<ListItem> listItems = parseJsonToList(response.toString());
                    emitter.onNext(listItems);
                    emitter.onComplete();
                } else {
                    Log.d("ListApiClient", "Error: HTTP " + responseCode);
                    emitter.onError(new Exception("Error: HTTP " + responseCode));
                }
            } catch (Exception e) {
                Log.d("ListApiClient", "Error: " + e.getMessage());
                emitter.onError(e);
            } finally {
                if (connection != null) {
                    connection.disconnect();
                }
            }
        });
    }

    private List<ListItem> parseJsonToList(String json) {
        Gson gson = new Gson();
        Type listType = new TypeToken<List<ListItem>>() {
        }.getType();
        return gson.fromJson(json, listType);
    }
}

File: MomentsKotlin/src/main/java/com/example/moments/data/api/ListApiClient.kt
package com.example.moments.data.api

import com.example.moments.data.model.ListItem
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.net.HttpURLConnection
import java.net.URL

private const val DOMAIN_URL = "192.168.31.246"
private const val TIME_OUT = 10000

class ListApiClient {
    suspend fun getList(): List<ListItem>? {
        var connection: HttpURLConnection? = null
        try {
            connection = URL("http://$DOMAIN_URL:3022").openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = TIME_OUT
            connection.readTimeout = TIME_OUT
            connection.connect()
            val responseCode = connection.responseCode
            if (responseCode == HttpURLConnection.HTTP_OK) {
                connection.inputStream.bufferedReader().use { reader ->
                    val json = reader.readText()
                    return parseJsonToList(json)
                }
            } else {
                throw Exception("Error loading data in ListApiClient: $responseCode")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        } finally {
            connection?.disconnect()
        }
    }

    private fun parseJsonToList(json: String): List<ListItem> {
        val listType = object : TypeToken<List<ListItem>>() {}.type
        return Gson().fromJson(json, listType)
    }
}